ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java

```diff
@@ -98,17 +99,99 @@ public String toString() {
         + '}';
   }
 
-  public KsqlFunction getFunction(final List<Schema.Type> paramTypes) {
-    final KsqlFunction function = functions.get(paramTypes);
+  public KsqlFunction getFunction(final List<Schema> paramTypes) {
+    final List<FunctionParameter> params = mapToFunctionParameter(paramTypes, true);
+    final KsqlFunction function = functions.get(params);
     if (function != null) {
       return function;
     }
 
+    if (paramTypes.stream().anyMatch(Objects::isNull)) {
+      return functions.entrySet()
+          .stream()
+          .filter(entry -> checkParamsMatch(entry.getKey(), params))
+          .map(Map.Entry::getValue)
+          .findFirst().orElseThrow(() -> createNoMatchingFunctionException(paramTypes));
+    }
+
+    throw createNoMatchingFunctionException(paramTypes);
+  }
+
+  private KsqlException createNoMatchingFunctionException(final List<Schema> paramTypes) {
     final String sqlParamTypes = paramTypes.stream()
-        .map(SchemaUtil::getSchemaTypeAsSqlType)
+        .map(schema -> schema == null
+            ? null
+            : SchemaUtil.getSchemaTypeAsSqlType(schema.type()))
         .collect(Collectors.joining(", ", "[", "]"));
 
-    throw new KsqlException("Function '" + metadata.getName()
+    return new KsqlException("Function '" + metadata.getName()
                             + "' does not accept parameters of types:" + sqlParamTypes);
   }
+
+  private static boolean checkParamsMatch(final List<FunctionParameter> functionArgTypes,
```

Comment:
If u change `suppliedParamTYpes` to being the `List<Schema>` I think the code ends up easier to understand as the match function becomes:

```
public boolean matches(final Schema schema) {
      if (schema == null) {
        return isOptional;
      }

      return type.equals(schema.type());
    }
```

i.e. match the function parameter to the schema of the invocation. 

(And I think you can then drop the 'includeNull' stuff).